simulation/behavior_tree_plugin/src/action_node.cpp

```diff
@@ -365,9 +369,142 @@ bool ActionNode::foundConflictingEntity(const std::vector<std::int64_t> & follow
   return false;
 }
 
-double ActionNode::calculateStopDistance(double deceleration) const
+auto ActionNode::calculateUpdatedEntityStatus(
+  double target_speed, const traffic_simulator_msgs::msg::DynamicConstraints & constraints) const
+  -> traffic_simulator_msgs::msg::EntityStatus
+{
+  const auto speed_planner =
+    traffic_simulator::longitudinal_speed_planning::LongitudinalSpeedPlanner(
+      step_time, entity_status.name);
+  const auto dynamics =
+    speed_planner.getDynamicStates(target_speed, constraints, getCurrentTwist(), getCurrentAccel());
+
+  double linear_jerk_new = std::get<2>(dynamics);
+  geometry_msgs::msg::Accel accel_new = std::get<1>(dynamics);
+  geometry_msgs::msg::Twist twist_new = std::get<0>(dynamics);
+
+  std::int64_t new_lanelet_id = entity_status.lanelet_pose.lanelet_id;
+  double new_s = entity_status.lanelet_pose.s +
+                 (twist_new.linear.x + getCurrentTwist().linear.x) / 2.0 * step_time;
+  if (new_s < 0) {
+    auto previous_lanelet_ids =
+      hdmap_utils->getPreviousLaneletIds(entity_status.lanelet_pose.lanelet_id);
+    new_lanelet_id = previous_lanelet_ids[0];
+    new_s = new_s + hdmap_utils->getLaneletLength(new_lanelet_id) - 0.01;
+    traffic_simulator_msgs::msg::EntityStatus entity_status_updated;
+    entity_status_updated.time = current_time + step_time;
+    entity_status_updated.lanelet_pose.lanelet_id = new_lanelet_id;
+    entity_status_updated.lanelet_pose.s = new_s;
+    entity_status_updated.lanelet_pose.offset = entity_status.lanelet_pose.offset;
+    entity_status_updated.lanelet_pose.rpy = entity_status.lanelet_pose.rpy;
+    entity_status_updated.action_status.twist = twist_new;
+    entity_status_updated.action_status.accel = accel_new;
+    entity_status_updated.action_status.linear_jerk = linear_jerk_new;
+    entity_status_updated.pose = hdmap_utils->toMapPose(entity_status_updated.lanelet_pose).pose;
+    return entity_status_updated;
+  } else {
+    bool calculation_success = false;
+    for (size_t i = 0; i < route_lanelets.size(); i++) {
+      if (route_lanelets[i] == entity_status.lanelet_pose.lanelet_id) {
+        double length = hdmap_utils->getLaneletLength(entity_status.lanelet_pose.lanelet_id);
+        calculation_success = true;
+        if (length < new_s) {
+          if (i != (route_lanelets.size() - 1)) {
+            new_s = new_s - length;
+            new_lanelet_id = route_lanelets[i + 1];
+            break;
+          } else {
+            new_s = new_s - length;
+            auto next_ids = hdmap_utils->getNextLaneletIds(route_lanelets[i]);
+            if (next_ids.empty()) {
+              return stopAtEndOfRoad();
+            }
+            new_lanelet_id = next_ids[0];
+            break;
+          }
+        }
+      }
+    }
+    if (!calculation_success) {
+      THROW_SIMULATION_ERROR(
+        "failed to calculate next status calculateUpdatedEntityStatus function");
+    }
```

Comment:
In your error message, state why it failed, not just "failed".

Reply:
Updated to more detailed description.